### ISSUE TITLE
Is there noProxy setting in proxy object?

### ISSUE THREAD

**Issue by tnguyen14** [mozilla/geckodriver#725]
For the `proxy` object (https://github.com/mozilla/geckodriver#proxy-object), is there a `noProxy` option to declare domains / IPs that will not go through the proxy, if `manual` is set for `proxyType` with setting for `httpProxy` and `sslProxy`?

**Comment by AutomatedTester** [comment]
WebDriver doesnt have this concept. It's always been an all or nothing approach. For this to be added it would need to be raised as an issue against https://github.com/w3c/webdriver

**Comment by tnguyen14** [comment]
So webdriver#913 has been closed, and according to this commit by @shs96c https://github.com/w3c/webdriver/commit/a90db4b119de8c77875647e30b611c1c49a4349b, there is support for `noProxy`? Can it be implemented in the geckodriver?

**Comment by whimboo** [comment]
I'm fairly sure that we will cover that via https://bugzilla.mozilla.org/show_bug.cgi?id=1370959

**Comment by whimboo** [comment]
We decided to add this feature before doing the huge refactoring as pointed out in my last comment. Reason is that it will give us more tests and lowers the risk for regressions. But you also get this feature earlier!

I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1387094 to get this implemented.

**Comment by estradino** [comment]
When is this due to be released?

**Comment by andreastt** [comment]
@estradino If you look at the bug @whimboo linked to, implementation has not started yet.

**Comment by estradino** [comment]
All right. Is there a way to go around that? Perhaps downgrade Firefox or use another version of the Geckodriver?

**Comment by andreastt** [comment]
You might be able to find a preference which achieves this effect.

**Comment by whimboo** [comment]
I started the implementation for webdriver (geckodriver), but then also have to get it implemented for Marionette (in Firefox). So it will be available in a couple of days but would require a new geckodriver release, which might be the one which takes a bit longer.

In the meantime just use the preference `network.proxy.no_proxies_on` to specify hosts which should not use the proxy when set to manual.

**Comment by whimboo** [comment]
This is all done now and will be available with Firefox 57 and the next release of geckodriver.

**Comment by tahaykACF** [comment]
I want to clear the `No Proxy for`list, so I tried the `selenium-webdriver 3.5.0`, `geckodriver-v0.19.0-linux64` with `Firefox Quantum 57.0b3`, `Firefox Quantum 57.0b4`, `FireFox Nightly 58.0a1` and `FireFox Nightly 58.0a1` using this proxy codes:

```javascript
var fproxy=proxy.manual({
				http: 'IP:PORT',
				bypass:[""]
				bypass:[]
                                bypass:undefined				
			});
```

for `bypass: undefined`, the `No Proxy for` list remains `localhost, 127.0.0.1`
for `bypass: []` and `bypass: [""]`, I get this error message: `InvalidArgumentError: Expected [object String] "" to be an Array`

**Comment by andreastt** [comment]
@tahaykACF See the README or the WebDriver specification for the accepted input.

**Comment by whimboo** [comment]
I have no idea what `bypass` should be here. If you are using a specific binding you should talk to the maintainers in how they are supporting the proxy capabilities. The above doesn't really tell me anything.

But what I noticed is that I missed to update the README file of geckodriver for proxy capabilities when working on the above mentioned bug. I filed [bug 1404891](https://bugzilla.mozilla.org/show_bug.cgi?id=1404891) to get this fixed.

Basically you need ´"noProxy": [list of hosts]` to list all exceptions.

**Comment by whimboo** [comment]
Oh and @tahaykACF thanks for noticing!

**Comment by tnguyen14** [comment]
Can this be re-opened?

According to https://github.com/SeleniumHQ/selenium/wiki/JsonWireProtocol#proxy-json-object, `noProxy` should be a string, not an array.

**Comment by whimboo** [comment]
No. In the [webdriver specification](https://w3c.github.io/webdriver/webdriver-spec.html#proxy) this is a list now. The binding you are using has to take care of it. See also https://github.com/SeleniumHQ/selenium/issues/5004.

**Comment by tnguyen14** [comment]
Ah ok. Do you know if there's an issue for the ChromeDriver to convert it to a list?

**Comment by whimboo** [comment]
You want to ask the Chrome team, and not us. :)

**Comment by tnguyen14** [comment]
Do you know if the webdriver spec for this was changed recently, or if it's been a list all along?

**Comment by whimboo** [comment]
Please lets really follow-up on https://github.com/SeleniumHQ/selenium/issues/5004.

There is nothing more to comment on this issue.
